rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+
+
+#### Schedule/Join Primitives
+
+The `strategy.run` API was initially developed for synchronous training. We propose a new pair of primitives to
+
+
+*   hide the details of load-balancing, fault tolerance and dynamic scheduling
+*   expose the non-blocking semantics to users.
+
+To enable scheduling a function on any worker, we recommend users create the same dataset, but may shuffle differently, on all workers via `strategy.experimental_distribute_datasets_from_function` API.
+
+
+```python
+class ParameterServerStrategyV2:
+  
+  def schedule(self, replica_fn, args=(), kwargs=()):
+    """Schedule the `replica_fn` on all replicas in a sync group (a worker).
+   
+    Schedule the `replica_fn` on all replicas in a sync group that is available,    
+    returns a future of PerReplica immediately if `function` has return values.
+
+    It implements at-least-once semantics for function execution. If a worker
+    fails, it will try to reschedule the function on another replica group or throw     
+    an exception to users. So this method assumes that function execution can be 
+    out of order and function inputs are shared between sync groups. 
+
+    We don't support the cases where `args` or `kwargs` are bound to a specific
+    sync group. We will consider supporting them in the future.
+
+    If there are barriers in `replica_fn`, it is users' responsibility to make
+    sure they won't cause deadlock.
+    """
+    pass
+  
+  def join(self):
+    """Wait until all scheduled functions are finished.
+
+    Raises an error if any of the functions fails to execute. In this case,
+    there is no guarantee that non-failing functions will complete.
+
+    When join() is being called, it is not allowed to call `schedule`.
+    """
+    pass
+
+  def done(self):
+    """Returns True if there is no pending functions to be executed."""
+    pass
+
+  def local_results(self, future_list):
+    """Get concrete values of the future list.
+    
+    Poisoned future objects will give `None`.
+    """
+    pass
+```
+
+
+#### Custom Training Loop
+
+To construct a custom training loop, users need to
+
+
+*   use `strategy.experimental_distribute_datasets_from_function` to create one dataset per worker. At this point, we recommend against using `strategy.experimental_distribute_dataset`.
+*   create models under `strategy.scope` so variables will be assigned to parameter servers.
+*   likewise, create a Keras metric object under `strategy.scope`. We expect the metric variables to be stored on parameter servers. Each worker, within their `replica_fn`, updates the metric states.
+*   use `strategy.schedule` to schedule the `replica_fn` on one remote worker. This `replica_fn` should take an iterator and perform forward and backward computation. This `strategy.schedule` will only schedule this `replica_fn` and returns one or several `Future` objects immediately.  
```

Comment:
What is a Future object? Is there an API contract, and is it a public API?

Reply:
A Future object possesses a field that may (once function returns result) or may not (when the function is only scheduled but not yet executed) have a concrete value. It's supposed to be used in conjunction with `local_result` with which the value of a Future is fetched.